MassSpectrometry/MsDataFile.cs

```diff
@@ -83,63 +83,175 @@ public static int TopNpeakHelper(ref double[] intensities, ref double[] mArray,
-                if (windowPeakIndexMinimum >= intensities.Length)
+            //set intensity of precursor ions to zero
+            for (int i = 0; i < mArray.Length; i++)
+            {
+                if (mArray[i] > (precursorMz - precursorDiscardRange) && mArray[i] < (precursorMz + precursorDiscardRange))
                 {
-                    break;
+                    intensities[i] = 0;
                 }
+            }
 
-                // determine the valid peaks given filtering conditions for this window
-                var windowIntensities = new double[windowPeakIndexMaximum - windowPeakIndexMinimum + 1];
-                Array.Copy(intensities, windowPeakIndexMinimum, windowIntensities, 0, windowIntensities.Length);
+            //filter peaks into discrete mass bins
+            MsDataFile.WindowModeHelper(ref intensities, ref mArray, firstFilter, (scanRangeMinMz - discreteMassBin / 2), (scanRangeMaxMz + discreteMassBin / 2));
 
-                var windowMzs = new double[windowPeakIndexMaximum - windowPeakIndexMinimum + 1];
-                Array.Copy(mArray, windowPeakIndexMinimum, windowMzs, 0, windowMzs.Length);
+            //normalize intensity in segments to max 50
+            FilteringParams secondFilter = new FilteringParams((int)Math.Ceiling((decimal)numberOfWindows / 10), null, 10, false, false);
+            MsDataFile.WindowModeHelper(ref intensities, ref mArray, secondFilter, (scanRangeMinMz - discreteMassBin / 2), (scanRangeMaxMz + discreteMassBin / 2), 50);
 
-                int numPeaksToTakeInThisWindow = TopNpeakHelper(ref windowIntensities, ref windowMzs, filteringParams);
+            for (int i = 0; i < mArray.Length; i++)
+            {
+                mArray[i] = Math.Round(mArray[i] / discreteMassBin, 0) * discreteMassBin;
+            }
 
-                // merge results of this window into the global results
-                intensityResults.AddRange(windowIntensities.Take(numPeaksToTakeInThisWindow));
-                mzResults.AddRange(windowMzs.Take(numPeaksToTakeInThisWindow));
+            //fill in missing mz values in array with zero intensity
+            List<double> allPossibleMzValues = new List<double>();
+            double start = Math.Round(scanRangeMinMz / discreteMassBin, 0) * discreteMassBin;
+            while (start < (scanRangeMaxMz + discreteMassBin / 2))
+            {
+                allPossibleMzValues.Add(start);
+                start += discreteMassBin;
+            }
 
-                // set up for the next window
-                windowPeakIndexMinimum = windowPeakIndexMaximum + 1;
-                windowPeakIndexMaximum = windowPeakIndexMinimum + windowSize;
+            //remove any that are already in the mArray
+            foreach (double remainingMzValue in mArray)
+            {
+                allPossibleMzValues.RemoveAll(x => x > (remainingMzValue - discreteMassBin / 2) && x < (remainingMzValue + discreteMassBin / 2));
```

Comment:
Isn't this wiping the entire bin?

Reply:
removes all mzs with intensity zero where there is a real non-zero mz intensity

Reply:
Can you add that as a comment?

Reply:
done